community/kernel/src/main/java/org/neo4j/kernel/impl/core/DelegatingTokenHolder.java

```diff
@@ -112,6 +119,79 @@ private synchronized int createToken( String name ) throws KernelException
         return id;
     }
 
+    @Override
+    public void getOrCreateIds( String[] names, int[] ids )
+    {
+        if ( names.length != ids.length )
+        {
+            throw new IllegalArgumentException( "Name and id arrays must have the same length." );
+        }
+        // Assume all tokens exist and try to resolve them. Break out on the first missing token.
+        boolean hasUnresolvedTokens = resolveIds( names, ids, ALWAYS_TRUE_INT );
+
+        if ( hasUnresolvedTokens )
+        {
+            createMissingTokens( names, ids );
+        }
+    }
+
+    private boolean resolveIds( String[] names, int[] ids, IntPredicate unresolvedIndexCheck )
+    {
+        boolean foundUnresolvable = false;
+        for ( int i = 0; i < ids.length; i++ )
+        {
+            Integer id = tokenCache.getId( names[i] );
+            if ( id != null )
+            {
+                ids[i] = id;
+            }
+            else
+            {
+                foundUnresolvable = true;
+                if ( unresolvedIndexCheck.test( i ) )
+                {
+                    // If the check returns `true`, it's a signal that we should stop early.
+                    break;
+                }
+            }
+        }
+        return foundUnresolvable;
+    }
+
+    private synchronized void createMissingTokens( String[] names, int[] ids )
+    {
+        // We redo the resolving under the lock, to make sure that these ids are really missing, and won't be
+        // created concurrently with us.
+        PrimitiveIntSet unresolvedIndexes = Primitive.intSet();
+        resolveIds( names, ids, i -> !unresolvedIndexes.add( i ) );
```

Comment:
A name duplicate would make resolveIds stop early, which would result in not all names being resolved/created. Problem?